src/Core/Framework/Util/Json.php

```diff
@@ -14,4 +14,28 @@ public static function encode(mixed $value, int $options = \JSON_UNESCAPED_UNICO
     {
         return (string) json_encode($value, $options);
     }
+
+    /**
+     * @throws \JsonException
+     *
+     * @return array<int, mixed>|array<string, mixed>|null
+     */
+    public static function decodeArrayOrObject(mixed $value): ?array
```

Comment:
I don't really understand why we need this method and why would you call it with an already decoded array? 

Can you explain to me why it is not enough to just do something like the following in the listener?:

```
$allowedIps = json_decode($salesChannelData['maintenanceIpWhitelist'] ?? '',  true, flags: \JSON_THROW_ON_ERROR) ?? [];
```

I guess I am missing something, sorry!

Reply:
done.

Reply:
@AydinHassan Sorry, I forgot why I added it, using an empty string with JSON_THROW_ON_ERROR produces an exception and I wanted to avoid the string type/empty string checks.

Reply:
however, I now check empty string and use an empty array in both corresponding methods.

Reply:
@AydinHassan I have thought about it some more and is there actually a reason why `JSON_THROW_ON_ERROR `is used? is this a guideline? 

because if we remove it your solution would work - we could skip empty etc. check - and if the json is invalid then a user who expected to have access will just not have access. same for the storefront maintenance resolver.

Reply:
Yeah it's a guideline because otherwise the error is hidden. Without the flag calling `json_last_error()` shows the same issue.

So we can use a default value of `[]` instead of empty string, or we can keep the method but I would rename it to just `decodeArray` and if it's invalid json or not an array, I would throw an exception, then calling code can decide wether it  cares or not. What do you think?

Reply:
@AydinHassan  it will never return an object. as `true` was used in json_decode. which means an object is returned as an associative array. 

I can throw an error at the beginning if we get an array or object as that should not happen at that point actually.

Reply:
actually I thinmk returning null if we do not get a string in that method makes the most sense.

Reply:
@AydinHassan I made some changes now. I now return null if no string is provided or the result is no array (which is always the case even for json objects).

Reply:
I mean that if the json passed is an object and not an array. I don't think we should blindly discard data. For a global utility like this, we should error if the passed json is not an array. 

So I would be happy if we inline this code and just ignore the result if it's not an array, but if we will introduce this global utility it should error on non array json input. Also, if we keep the global utility, please rename it to `decodeArray`.

Reply:
decodeArray is wrong though. an array in json is only `["abc", "abc2"]` but `{"abc": "abc"}`will be decoded correctly as well and that is an object. which is why I still think decodeArrayOrObject is correct. but I am fine with renaming it, but I think that would be confusing.

Regarding its functionality, the following should be the case:

- `null` should return `null`.
- empty string should return `null`.
- non string value should throw an exception.
- result of json decode not being an array, should throw an exception.
- result of json decode being an array, return it.

correct?

Reply:
I would only accept a string, so type hint `string`.

Empty string should return empty array (we don't discard anything here).

Throw an exception if the result of decode is not an array.

decodeArray with `array_is_list` sounds good to me. Very clear API.

Reply:
@AydinHassan I have now adjusted the json decodeArray accordingly. can you check if the implementaiton now is as expected. 

only not sure about what should happen if the value of the array is an object, should it decode those to objects or to an associative array. depending on the expected result, those two test cases would be different:
\Shopware\Tests\Unit\Core\Framework\Util\JsonTest::testDecodeArrayDecodesObjectWithSequentialNumericIndicesAsArray
\Shopware\Tests\Unit\Core\Framework\Util\JsonTest::testDecodeArrayWithObjectsAsArrayListWithAssociativeArrays

Reply:
I think how you have it now is correct, we should only care about top level list, eg an list of objects is legitimate, and the first example is also fine, we don't discard any data by converting the json object to an array as it is sequential.

Reply:
ok. I think then this should be it now?

Reply:
yep, thanks for the collaboration! I will reimport in to our Gitlab again and will let you know when it's merged

Reply:
@AydinHassan I just noticed that phpstan has one error regarding domain exception and not using UnexpectedValueException directly. However, inside this utility it makes little sense to create a HTTP based domain exception, should I still create one and if so where?

Reply:
I already took care of it :) - it's in internal review now